base/linalg/dense.jl

```diff
@@ -309,17 +309,65 @@ kron(a::AbstractVector, b::AbstractVector)=vec(kron(reshape(a,length(a),1),resha
 kron(a::AbstractMatrix, b::AbstractVector)=kron(a,reshape(b,length(b),1))
 kron(a::AbstractVector, b::AbstractMatrix)=kron(reshape(a,length(a),1),b)
 
-^(A::AbstractMatrix, p::Integer) = p < 0 ? inv(A^-p) : Base.power_by_squaring(A,p)
-
-function ^(A::AbstractMatrix, p::Number)
+# Matrix power
+^{T}(A::AbstractMatrix{T}, p::Integer) = p < 0 ? Base.power_by_squaring(inv(A),-Integer(p)) : Base.power_by_squaring(A,Integer(p))
+function ^{T}(A::AbstractMatrix{T}, p::Real)
+    n = checksquare(A)
+    # Quick return if A is diagonal
+    if isdiag(A)
+        retmat = copy(A)
+        for i in 1:n
+            retmat[i, i] = retmat[i, i] ^ p
+        end
+        return retmat
+    end
+    # For integer powers, use repeated squaring
     if isinteger(p)
-        return A^Integer(real(p))
+        TT = Base.promote_op(^, eltype(A), typeof(p))
+        return (TT == eltype(A) ? A : copy!(similar(A, TT), A))^Integer(p)
+    end
+
+    # If possible, use diagonalization
+    if issymmetric(A) && T <: Real
+        return full(Symmetric(A)^p)
+    end
+    if ishermitian(A)
+        return full(Hermitian(A)^p)
+    end
+
+    # Otherwise, use Schur decomposition
+    if istriu(A)
+        #Integer part
+        retmat = full(A) ^ floor(p)
+        #Real part
+        retmat = retmat * full(powm(UpperTriangular(float.(A)), real(p - floor(p))))
+        d = diag(A)
+    else
+        S,Q,d = schur(complex(A))
+        #Integer part
+        R = full(S) ^ floor(p)
+        #Real part
+        R = R * full(powm(UpperTriangular(float.(S)), real(p - floor(p))))
+        retmat = Q * R * Q'
+    end
+
+    # Check whether the matrix has nonpositive real eigs
+    np_real_eigs = false
+    for i = 1:n
+        if imag(d[i]) < eps() && real(d[i]) <= 0
+            np_real_eigs = true
+            break
+        end
+    end
+    if np_real_eigs
```

Comment:
Unconditional warnings like this are typically annoying because they end up being thrown in loops and spams the terminal. Isn't it enough to document what happens and then do that thing silently?